### ISSUE TITLE
feat(model): paged exact-span eviction with FP8/INT8 KV quantization and f32 Kraw shadow tracking

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12321]
## Context & Motivation

In `internal/model/paged_evict.go` and `docs/notes/EVICT-ON-PAGED-KV-DESIGN-2026-06-28.md` §6, `NewPagedKVPoolWithRaw` proved that fak's signature bit-exact middle-span eviction (`Evict`) cleanly survives block-paged KV allocation via a 3-plane pool ($K, V, K_{\text{raw}}$). Shifted survivors have their post-RoPE $K$ re-derived in a single rotation from pre-RoPE $K_{\text{raw}}$ at their new logical indices ($\max|\Delta| = 0$), avoiding the numerical drift of composed rotations that flips greedy tokens.

However, in-kernel float32 KV cache requires 3 rows × 4 bytes = 12 bytes per element per layer ($K, V, K_{\text{raw}}$), which severely limits context capacity and multi-tenant batch sizes. Quantized KV caches (FP8 E4M3/E5M2 or INT8 Q8_0) reduce attention read memory by 2×–4×, but lossy quantized keys cannot be re-rotated without substantial numerical drift ($\sim 10^{-3}$ to $10^{-2}$).

As documented in #1047 (`internal/compute/kvprecision.go`) and the design note, the robust solution is an **f32 $K_{\text{raw}}$ shadow plane**: the two attended planes ($K$ and $V$) are quantized to 8-bit for compact memory and high-throughput attention, while an uncompressed float32 $K_{\text{raw}}$ shadow plane is retained to power bit-exact single-rotation re-RoPE during exact-span eviction.

---

## For / Problem / Today / Better because

- **For:** Autonomous agent serving runtimes executing dynamic context quarantine, speculative branching with CoW rollback, and memory-constrained multi-turn agent serving where middle-span eviction (`Evict`) must preserve bit-exact generation.
- **Problem:** Full F32 paged KV caches (`NewPagedKVPoolWithRaw`) cost 12 bytes/element/layer ($K + V + K_{\text{raw}}$), severely restricting KV cache density and maximum concurrent sequences. Standard quantized KV caches (FP8/INT8) compress attention tensors to 1–2 bytes/element, but cannot perform exact-span eviction: composing two lossy rotations or rotating from quantized $K$ induces numerical drift ($\sim 10^{-3}$ to $10^{-2}$), flipping greedy tokens and violating the `== never-saw-it` invariant.
- **Today:** `paged_evict.go` implements bit-exact paged eviction using an unquantized 3-plane F32 layout ($K, V, K_{\text{raw}}$). Meanwhile, `internal/compute/kvprecision.go` (#1047) defines precision tiers (F32 and Q8 mixed), but `PagedKVPool` only supports uniform `[]float32` planes. Quantized pools currently have no way to retain pre-RoPE keys, forcing quantized serving to degrade to whole-prefix reset (`SupportsExactSpan = false`), which wipes the entire prefix and triggers expensive recomputations.
- **Better because against next-best alternatives:**
  - *vs. Whole-prefix flush / Radix truncation (vLLM `reset_prefix_cache` / SGLang `flush_cache`):* vLLM/SGLang drop all downstream context when an agent tool output or prompt injection is invalidated. fak preserves valid downstream context by splicing out only the exact span $[from, from+n)$ without recomputing prefix tokens.
  - *vs. Uniform FP8/INT8 without shadow:* Rotating from lossy 8-bit quantized $K$ accumulates rotational error and flips greedy tokens, breaking the deterministic trust floor.
  - *vs. Unquantized 3-plane F32 paged KV:* The 4-plane layout ($K_{\text{quant}}, V_{\text{quant}}, \text{scales}, K_{\text{raw}}^{\text{f32}}$) reduces per-token memory from 12 bytes/elem/layer down to 6 bytes/elem/layer (a ~2× memory density gain), doubling fittable context while maintaining $\max|\Delta| = 0$ bit-exact eviction.

---

### 1. Quantized Block Layout (4 Planes)
A quantized paged pool block must organize four distinct functional planes per token per layer:
1. **Plane 0 ($K_{\text{quant}}$):** 1 byte/element (FP8 E4M3 / E5M2 or INT8 symmetric).
2. **Plane 1 ($V_{\text{quant}}$):** 1 byte/element (FP8 E4M3 / E5M2 or INT8 symmetric).
3. **Plane 2 ($S_{\text{quant}}$):** Quantization scale factors (and optional zero-points/biases) stored as float32/float16 (e.g., per-head or per-32/128-element tile scale).
4. **Plane 3 ($K_{\text{raw}}$ Shadow):** Unquantized float32 (4 bytes/element) storing the pre-RoPE Key vector.

`PagedKVPool` storage representation must be updated to accommodate heterogeneous types rather than a homogeneous `[][]float32`:
- Dedicated byte slabs for quantized tensor storage ($K_{\text{quant}}, V_{\text{quant}}$).
- Separate or structured float32 storage for scales and the $K_{\text{raw}}$ shadow plane.
- Zero-copy indexing: block offset calculation `slotQuant(layer, plane, tok)` and `slotRaw(layer, tok)` must maintain memory alignment for vector loads.

### 2. Single-Rotation Re-RoPE from Shadow
During `Evict(from, n, cfg)`:
- Unmoved survivors ($op < from$): retain their $K_{\text{quant}}$, $V_{\text{quant}}$, scales, and $K_{\text{raw}}$ untouched.
- Shifted survivors ($op \ge from + n$): shifted to new logical position $i = op - n$.
  - $V_{\text{quant}}$ and $K_{\text{raw}}$ move verbatim ($V$ is never rotated; $K_{\text{raw}}$ is position-independent).
  - Post-RoPE $K$ is re-derived in full float32 via `reropeRowFromRaw(cfg, l, i, kraw[l])` in a **single rotation** at logical index $i$.
  - Re-derived float32 $K$ is freshly quantized into $K_{\text{quant}}$ using the pool's quantization kernel, updating $S_{\text{quant}}$ accordingly.
  - Guarantees that $K_{\text{quant}}$ matches the exact quantization of a sequence prefilled directly at position $i$ without rotational drift accumulation.

### 3. Copy-on-Write (CoW) Block Splitting
- Because re-RoPE and compaction mutate block storage, sequences sharing physical blocks with un-evicted forks (created via `Fork()`) must split safely.
- `Evict()` must rebuild shifted survivors into freshly allocated private blocks (or invoke `ensureOwned(li)` on target blocks), decrementing references on the shared parent blocks.
- Parent sequences maintain reference counts and retain their original post-RoPE $K_{\text{quant}}$ and $K_{\text{raw}}$ bytes completely untouched.

---

## Edge Cases & Failure Modes

1. **Memory Budget & Overhead Accounting:**
   - Pure FP8/INT8 KV without shadow is 2 bytes/token/layer ($1B_K + 1B_V$ + marginal scale).
   - With f32 $K_{\text{raw}}$ shadow: $1B_K + 1B_V + 4B_{K_{\text{raw}}} = 6B$ + scale/token/layer, compared to $4B_K + 4B_V + 4B_{K_{\text{raw}}} = 12B$ in F32.
   - The memory planner (`internal/compute/kvprecision.go`) and internal fragmentation reporter (`OverheadRatio()`) must account for the mixed 6-byte footprint (`StorageLabel: "mixed"`).
2. **Dimension & Memory Alignment:**
   - Stride per layer = $\text{NumKVHeads} \times \text{HeadDim}$.
   - Block allocations must satisfy SIMD/GPU vector alignment (16-byte / 32-byte / 64-byte boundaries) for both 1-byte quantized rows and 4-byte float32 shadow rows.
3. **Dequantization Overhead vs Attention Throughput:**
   - Attention read paths (`GatherK`, `GatherV`, or native compute backends) must either support fused quantized attention or fast on-the-fly vector dequantization.
   - Re-quantization during eviction is an infrequent control-plane event and must avoid dynamic heap allocations during re-RoPE loops.
4. **Fallback for Pools Without Shadow:**
   - If a pool is initialized in pure quantized mode without the $K_{\text{raw}}$ shadow plane (to maximize memory savings when eviction is not required), `SupportsRaw()` returns `false`.
   - Any call to `Evict()` must fail closed (or gracefully degrade via `EvictGoverned` reporting `att.ExactSpanSupported = false` and `att.Degraded = true`), refusing to perform lossy composed rotations.

---

## Deterministic Witness Command

```bash
go test -v ./internal/model -run "TestPagedEvictQuantizedKrawShadow"
```

The witness test must assert:
1. Shifted survivors re-RoPE'd from the f32 $K_{\text{raw}}$ shadow match a freshly prefilled reference sequence under the same quantizer.
2. Unquantized float32 intermediate keys re-derived during eviction remain bit-identical ($\max|\Delta| = 0$) to `KVCache.Evict`.
3. CoW isolation: a parent sequence sharing blocks via `Fork()` is byte-for-byte unchanged when a child sequence executes middle-span eviction.
4. End-to-end logits: next-token decode logits from the evicted quantized paged sequence match the logits of a freshly prefilled identical sequence within the expected quantization error bound.

---

## First Checkable Step / Definition of Done

- **First Checkable Step:**
  - Define `PagedKVQuantMode` (FP8 E4M3, INT8 Q8_0) in `internal/model`.
  - Extend `PagedKVPool` to support the 4-plane block configuration (`NewPagedKVPoolQuantizedWithRaw(cfg, blockTokens, quantMode)`).
  - Implement `AppendQuantizedRaw` and verify round-trip retrieval via `GatherKraw`.
- **Definition of Done:**
  - [ ] 4-plane quantized block memory layout implemented in `internal/model/pagedkv.go`.
  - [ ] Single-rotation re-RoPE from f32 $K_{\text{raw}}$ shadow and subsequent requantization implemented in `internal/model/paged_evict.go`.
  - [ ] CoW block splitting verified on forked sequences during eviction.
  - [ ] Deterministic witness test `TestPagedEvictQuantizedKrawShadow` passes in CI.
  - [ ] Capacity and memory plan calculations in `internal/compute/kvprecision.go` aligned with the 4-plane mixed layout.
